lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp

```diff
@@ -2241,22 +2240,70 @@ void HdVP2Mesh::_UpdateDrawItem(
                 stateToCommit._ufeIdentifiers.length()
                 == stateToCommit._instanceTransforms.length());
 #endif
-            if (stateToCommit._instanceTransforms.length() == 0)
-                instancerWithNoInstances = true;
-
-            // instancer with no instances means nothing to draw. Disable
-            // the render item if it is not already disabled
-            if (instancerWithNoInstances) {
-                if (drawItemData._enabled) {
-                    drawItemData._enabled = false;
-                    _delegate->GetVP2ResourceRegistry().EnqueueCommit(
-                        [renderItem]() { renderItem->enable(false); });
-                }
-                // skip the rest of the update because the MRenderItem is not
-                // enabled
-                return;
+        }
+
+        // instancer with no instances means nothing to draw. Disable
+        // the render item if it is not already disabled
+        if (!stateToCommit._instanceTransforms
+            || stateToCommit._instanceTransforms->length() == 0) {
+            if (drawItemData._enabled) {
+                drawItemData._enabled = false;
+                _delegate->GetVP2ResourceRegistry().EnqueueCommit(
+                    [renderItem]() { renderItem->enable(false); });
+            }
+            // skip the rest of the update because the MRenderItem is not
+            // enabled
+            return;
+        }
+
+        // compare the new _instanceTransforms on stateToCommit to
+        // the existing instance transforms (if any) on drawItemData
+        bool instanceTransformsChanged = static_cast<bool>(stateToCommit._instanceTransforms)
+            ? !static_cast<bool>(drawItemData._instanceTransforms)
+            : static_cast<bool>(drawItemData._instanceTransforms);
+        if (stateToCommit._instanceTransforms && drawItemData._instanceTransforms) {
+            instanceTransformsChanged
+                = (stateToCommit._instanceTransforms->length()
+                   != drawItemData._instanceTransforms->length());
+            for (unsigned int index = 0;
+                 index < stateToCommit._instanceTransforms->length() && !instanceTransformsChanged;
+                 index++) {
+                instanceTransformsChanged
+                    = ((*stateToCommit._instanceTransforms)[index]
+                       != (*drawItemData._instanceTransforms)[index]);
```

Comment:
As painfully slow as this is, setting the data redundantly into Maya is much slower because Maya rebuilding the instanced render item in VP2 (interacting with the graphics driver). Here we're doing work on the CPU, and this code is threaded so the scalability is still okay.

Reply:
@williamkrick can we use memcmp to compare two buffers here? Theoretically it may be faster.